[clang-format] Allow array alignment on non-rectangular arrays

bdunkin opened the pull request:
This change rewrites the implementation of `AlignArrayOfStructures` to address current limitations and formatting bugs. It now also handles non-rectangular arrays (arrays of arrays where the inner arrays do not all have the same number of elements).

There were a couple significant sources of formatting bugs in the previous implementation:

- Modifying newlines during alignment instead of during the line break optimization pass. This caused many unexpected formatings because it overwrote the result of the line break optimization pass, ignoring almost all formatting rules.
- Modifying spaces before computing alignment information. This caused cases like a single row to change format, even when there was no alignment to be made.
- Not keeping all the bookkeeping around token columns up to date when alignment was made. This was not unique to `AlignArrayOfStructures`, and this change fixes all cases. This meant trailing comments, and escaped newlines would not always align correctly.
- Incorrect handling of tabs for alignment within a line, as opposed to the beginning of a line.

All of these issues have been fixed, however to do so there is a significant backwards incompatible change to how rows with line breaks are handled. Previously the columns before and after the line break would attempt to be aligned with other rows. With this change, all elements in such a row will instead be placed on their own line, and no alignment will be attempted, similar to if a trailing comma is added to a list. Trying to align columns in this case had too many edge cases. Some could not reasonably be addressed without knowing alignment information during line break optimization, which is not possible because it would require knowing the formatting of future tokens. See the changed tests for examples of the difference.

Fixed Issues (some are fixed only in that they no longer produce wacky alignment, rather than what the user wished it produced):
https://github.com/llvm/llvm-project/issues/138151 https://github.com/llvm/llvm-project/issues/85937 https://github.com/llvm/llvm-project/issues/76803 https://github.com/llvm/llvm-project/issues/76803 https://github.com/llvm/llvm-project/issues/76717 https://github.com/llvm/llvm-project/issues/55477 https://github.com/llvm/llvm-project/issues/55154 https://github.com/llvm/llvm-project/issues/53442 https://github.com/llvm/llvm-project/issues/55485 https://github.com/llvm/llvm-project/issues/55477 https://github.com/llvm/llvm-project/issues/54354 https://github.com/llvm/llvm-project/issues/51766 https://github.com/llvm/llvm-project/issues/86439 https://github.com/llvm/llvm-project/issues/142072 https://github.com/llvm/llvm-project/issues/89721

Attribution Note - I have been authorized to contribute this change on behalf of my company: ArenaNet LLC

bdunkin committed:
Rewrite AlignArrayOfStructures implementation to allow non-rectangular arrays,
and fix numerous formatting bugs.

bdunkin committed:
Allow array alignment to happen on non-rectangular arrays

bdunkin commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -27965,39 +27995,39 @@ TEST_F(FormatTest, AlignArrayOfStructuresRightAlignmentNonSquare) {
   verifyFormat("struct test demo[] = {\n"
-               "    {1, 2, 3, 4, 5},\n"
-               "    {3, 4, 5},\n"
+               "    {1, 2, 3, 4,  5        },\n"
```
Note: the 5 was actually _not_ properly aligned in this test before as the column in the last row has two character.

HazardyKnusperkeks commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -27901,39 +27901,39 @@ TEST_F(FormatTest, AlignArrayOfStructuresLeftAlignmentNonSquare) {
   // also act as regression tests for when it is properly fixed
   verifyFormat("struct test demo[] = {\n"
-               "    {1, 2},\n"
+               "    {1, 2   },\n"
```
That's a solid point. I personally think it looks nicer, just don't know about changing the formatting (and thus the tests).
Let's see what others think.

mydeveloperday reviewed (changes requested):
You can't change existing tests, you need an option if you want to pad with spaces

HazardyKnusperkeks commented:
> > You can't change existing tests, you need an option if you want to pad with spaces
> 
> It is unclear to me if this should be considered a bug fix, or a new feature. Surely requiring new options are not required for every bug fix.
> 

It boils down to the question if this is seen as a bug or not. As already stated, in my opinion it looks nicer with the padding, but I have no opinion on wether it is currently a bug or not.

bdunkin commented:
I will gladly add a new option, if that is deemed necessary. I just want to make sure that's the direction I should go. My experience has been it is very hard to take back an option once it has been released, so they should be added deliberately, and with careful consideration.

bdunkin commented:
Which existing bugs need to be fixed? I tried to not change the underlying code too much, but if a redesign is on the table, does that mean backwards compatibility is a lesser concern?

owenca commented:
> Which existing bugs need to be fixed? 

https://github.com/llvm/llvm-project/issues?q=is%3Aopen%20is%3Aissue%20in%3Atitle%20AlignArrayOfStructures. There may be more, e.g. when the option name is not in the issue title.

> I tried to not change the underlying code too much, but if a redesign is on the table, does that mean backwards compatibility is a lesser concern?

Yes, if the new design fixes existing bugs or clearly improve the alignment.

mydeveloperday commented:
I couldn't approve this change with the test changing so... we'll just get beaten by people saying "you changed my code"

bdunkin commented on clang/lib/Format/WhitespaceManager.cpp:
```diff
@@ -278,6 +278,33 @@ void WhitespaceManager::calculateLineBreakInformation() {
+// subsequent tokens so that trailing comments and escaped newlines can be
+// aligned properly
+static void
+SetChangeSpaces(unsigned Start, unsigned Spaces,
```
Updating `Spaces`, without also modifying `PreviousEndOfTokenColumn` for the next token means the spaces are not considered when aligning trailing comments or preprocessor line breaks. I changed all places in this file that modified `Spaces` directly to use this function instead.

bdunkin commented on clang/lib/Format/WhitespaceManager.cpp:
```diff
@@ -1318,281 +1343,253 @@ void WhitespaceManager::alignArrayInitializers(unsigned Start, unsigned End) {
   return CellWidth;
 }
 
-void WhitespaceManager::alignToStartOfCell(unsigned Start, unsigned End) {
```
This function is no longer used.

bdunkin commented on clang/lib/Format/WhitespaceManager.cpp:
```diff
@@ -1617,37 +1618,39 @@ void WhitespaceManager::generateChanges() {
-      // between 2 tokens and another non-empty range at the start of the second
-      // token.  We didn't implement logic to combine replacements for 2
-      // consecutive source ranges into a single replacement, because the
+      // between 2 tokens and another non-empty range at the start of the
```
I am unsure why this comment has been different line wrapping now, but this is the current output of clang-format, so...

bdunkin commented on clang/lib/Format/WhitespaceManager.h:
```diff
@@ -274,76 +268,17 @@ class WhitespaceManager {
   CellDescriptions getCells(unsigned Start, unsigned End);
 
-  /// Does this \p Cell contain a split element?
-  static bool isSplitCell(const CellDescription &Cell);
```
These functions are no longer used.

bdunkin commented on clang/lib/Format/WhitespaceManager.cpp:
```diff
@@ -118,9 +118,9 @@ const tooling::Replacements &WhitespaceManager::generateReplacements() {
     alignConsecutiveTableGenDefinitions();
   }
   alignChainedConditionals();
+  alignArrayInitializers();
```
Aligning elements in arrays can changed the length of a line, so we must do it before trailing comments and escaped newlines, otherwise they will not be working with the final length of the line.

bdunkin commented:
@owenca @mydeveloperday @HazardyKnusperkeks I have updated this change to address a significant number of formatting issues related to this option. It is almost a full rewrite of the feature, but I believe it now plays much more nicely with other formatting rules. Rows with line breaks in them can't reasonably have column alignment, and I made a choice to handle them as if they were a list with a trailing comma, but I am certainly open to other handling if there is a better suggestion.

HazardyKnusperkeks reviewed (commented):
Haven't looked over everything. But we still have the issue of changed tests (and I think more then before?), before there isn't an agreement, a thorough review is maybe just a waste of time.

bdunkin commented:
@owenca just a friendly bump on this review. The main question is currently whether changing tests is acceptable.

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -22118,11 +22133,14 @@ TEST_F(FormatTest, CatchAlignArrayOfStructuresRightAlignment) {
-      "     \"just world, ought to be split over multiple lines\"},\n"
-      "    {-1, 93463,                                  \"world\"},\n"
-      "    {56,     5,                                     \"!!\"}\n"
+      "    {\n"
+      "        7,\n"
+      "        23,\n"
+      "        \"hello world i am a very long line that really, in any\"\n"
+      "        \"just world, ought to be split over multiple lines\"\n"
+      "    },\n"
```
This is worse than before IMO.

owenca commented:
> @owenca just a friendly bump on this review. The main question is currently whether changing tests is acceptable.

See https://github.com/llvm/llvm-project/pull/143781#issuecomment-3071784802. For example, https://github.com/llvm/llvm-project/pull/143781#discussion_r2312911081 is worse than before.

bdunkin commented:
I have updated the PR description to cover what I've changed here.

To address directly the test you point out: The alignment code previously would add line breaks to long string literals that had been aligned so they crossed the column limit. The resulting format ignored other formatting rules, had edge cases that were not considered, and overwrote the result of the line break optimization pass where those edge cases _were_ considered. This caused a lot of bad formatting because it created formatting that the line break optimization would not have chosen. I could not find a satisfactory way to address the edge cases so I change the alignment code to not add or remove line breaks.

All line breaks are now added _only_ in the line break optimization pass, and alignment has to work within them. This means that any row that does not fit on a line needs a rule that can be applied at line break optimization time. I chose for them to be formatted like lists with a trailing comma (one element per line), but I have no strong attachment to that choice if someone comes up with an alternative.

owenca commented:
There are compiler warnings, e.g.:
```
clang/unittests/Format/FormatTest.cpp:22507:25: warning: unknown escape sequence '\d' [-Wunknown-escape-sequence]
 22507 |                "    { \"\data\",   Code::X },\n"
       |                         ^~
```

HazardyKnusperkeks commented:
Maybe @sstwcw has some opinion on that. Otherwise we could just go forward next week.

sstwcw commented:
Are you still working on this patch?

bdunkin committed:
Rebase on to latest main, remove accidental '\d' characters, and replace new tests' style options that have been changed in main.

sstwcw commented:
I don't see a major problem.  The reviewers are usually less patient.

bdunkin commented:
I rebased on latest main to address merge conflicts. There were no changes to the logic needed, it was just an unrelated bug fix that happened to overlap with an area that I had also made changes to.

owenca commented:
It seems that we are waiting for the following:
- @mydeveloper to confirm that his requested change was addressed
- @HazardyKnusperkeks, @sstwcw, and @owenca to mark their review comments as resolved
- an approval from one of the reviewers

mydeveloperday reviewed (changes requested):
We can change all these tests, we might even think we are doing the right thing now, the problem is users will complain they upgrade and the default behavior as changed

mydeveloperday commented:
> **We can't change** all these tests, we might even think we are doing the right thing now, the problem is users will complain they upgrade and the default behavior as changed

sorry typo
